proposals/nnnn-actors.md

```diff
@@ -140,6 +140,265 @@ extension BankAccount {
+    async let cash = checkCash()
+    let available = await cash - debt.amount
+    
+    if await requested <= available { 
+      amount -= requested
+      await friend.send(cash: requested) // ✅        
+    }  
+  }
+
+  func looseWallet() async { 
+    amount = 0
+  }
+}
+```
+
+This example composes well; Calls to an actor's `self` are allowed to pass through as usual, and _replies_ from other actors are also accepted. What can _not_ happen with non-reentrant actors is the `looseWallet()` function being randomly triggered while we are attempting to lend our friend some cash -- this would have been an external call into the actor, which our non-reentrancy rule would prevent.
+
+So even such snippet (under non-reentrancy rules):
+
+
+```swift
+await wallet.lendFriendSomeCash()
+await wallet.looseWallet() 
+```
+
+would be correct. Even if our friend takes 10 minutes to reply to `howMuchDoYouNeed`, we are being patient with them and wait with processing the next _external_ message (that will cause us to loose our wallet), until after we are done lending our friend some cash.
+
+Under reentrant rules, the above code could be unsafe, changing our wallet's balance to zero right before we are about to decrement it (!).
+
+#### Proposal: Opt-in reentrancy
+
+We could amend our model with the following changes:
+
+- actors are **non-**reentrant by default
+  - external calls from other actors are performed "one by one," they cannot automatically "jump in front of the queue"
+- actor-isolated functions may be annotated with `@alwaysInterleave`, which enables given function to run "interleaved" with other actor functions. This enables high-priority request handling and similar patterns.
+- actors which want all their functions to be able to interleave each other, may annotate their definition using `@reentrant`, which causes all functions isolated to this actor to be `@alwaysInterleaving` 
+
+> Alternate spelling proposals follow below the example.
+
+The rationale to make actors non-reentrant by default is clear: it is what feels natural and what actually enables "write as if synchronous code in an actor, and it just works" style of programming.
+
+However, there are valid and important cases where we _do_ want to enable reentrant calls - e.g. a high priority message changing how we are processing a long running task inside an actor:
+
+```swift
+actor class ImageDownloader { 
+  var bestImages: [Image]
+  var currentBest: Image?
+  
+  func downloadAndPickBest(n: Int, urls: [URL]) async -> [Image] {
+    for url in urls {
+      let image = await download(url)
+      bestImages.append(image)
+      let ranking = await rank(image) 
+      if ranking.isBest { 
+        bestImage = image  
+      }
+      // ... more things to compute only the "best n" etc...
+    }
+
+    return images
+  }
+  
+  @alwaysInterleave // ✅ may be invoked at any point in time 
+  func currentBest() async -> Image { 
+    currentBest
+  } 
```

Comment:
I don't see how this is consistent with what you're saying.  If programs can just add an `@alwaysInterleave` method to an actor, then every other async function on that actor still has to worry about interleaving.  The idea that's consistent with your argument is that async functions have to consent with interleaving.

Reply:
Hm yes... this allows for throwing in a wrench and "always interleave anyway" kind of intended for readonly queries or cancellation-like things mostly.

Your proposal makes sense, I guess it would be a `@reentrant func` then, which allows to be interleaved with.

Reply:
I rephrased the proposal to focus on specific functions opting into being reentrant, or the full actor opting in via `@reentrant` annotations.

I kept a mention of `@interleave(readOnly/unsafe)` but am not sure if we need it or not.